### ISSUE TITLE
Zero-byte read on sockets sometimes leads to receiving the same data repeatedly on macOS 

### ISSUE THREAD

**Issue by JunTaoLuo** [dotnet/corefx#31766]
Hey @stephentoub, I ran into what appears to be a socket bug while introducing a zero-byte read as part of https://github.com/aspnet/KestrelHttpServer/pull/2502 (issue: https://github.com/aspnet/KestrelHttpServer/issues/2334). 

The read loop looks like the following:
```c#
            while (true)
            {

// this section is newly added
                _trace.LogDebug("Waiting for data");
                // Wait for data before allocating a buffer.
                var waitedBytes = await _receiver.WaitForDataAsync();
                _trace.LogDebug($"Waited for {waitedBytes} bytes of data");
// up to this point

                // Ensure we have some reasonable amount of buffer space
                var buffer = Input.GetMemory(MinAllocBufferSize);

                _trace.LogDebug($"Receiving data with a buffer at {buffer.GetArray().Offset} with count {buffer.GetArray().Count}");
                var bytesReceived = await _receiver.ReceiveAsync(buffer);

                _trace.LogDebug($"Received {bytesReceived} bytes of data");
...
            }
```
Where the implementation of `WaitForDataAsync` and `ReceiveAsync` are as follows:
```c#        
        public SocketAwaitableEventArgs WaitForDataAsync()
        {
            _awaitableEventArgs.SetBuffer(Array.Empty<byte>(), 0, 0);

            if (!_socket.ReceiveAsync(_awaitableEventArgs))
            {
                _awaitableEventArgs.Complete();
            }

            return _awaitableEventArgs;
        }

        public SocketAwaitableEventArgs ReceiveAsync(Memory<byte> buffer)
        {
#if NETCOREAPP2_1
            _awaitableEventArgs.SetBuffer(buffer);
#elif NETSTANDARD2_0
            var segment = buffer.GetArray();

            _awaitableEventArgs.SetBuffer(segment.Array, segment.Offset, segment.Count);
#else
#error TFMs need to be updated
#endif
            if (!_socket.ReceiveAsync(_awaitableEventArgs))
            {
                _awaitableEventArgs.Complete();
            }

            return _awaitableEventArgs;
        }
```

It seems like introducing the zero-byte read (in order to reduce memory used by idle connections) can occasionally cause the same data to be received repeatedly. For example, in normal conditions, I would expect to read new data for every receive:
```
[0.020s] [TestLifetime] [Information] Starting test "DoesNotOverrideDirectConfigurationWithIServerAddressesFeature_IfPreferHostingUrlsFalse" at "2018-08-14T17:57:17"
[0.119s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting starting
[0.332s] [Microsoft.AspNetCore.Server.Kestrel] [Warning] Overriding address(es) '"http://127.0.0.1:0"'. Binding to endpoints defined in "UseKestrel()" instead.
[0.362s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting started
[0.362s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Loaded hosting startup assembly "Sockets.BindTests"
[0.425s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLG21TG3M129"" started.
[0.448s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.461s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
[0.468s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Receiving data with a buffer at 123288 with count 4096
[0.469s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Received 169 bytes of data
[0.473s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.483s] [LoggingConnectionAdapter] [Debug] ReadAsync[5] 16 03 02 00 A4 
¤
[0.485s] [LoggingConnectionAdapter] [Debug] ReadAsync[164] 01 00 00 A0 03 03 5B 73 17 FE 60 2E 15 8D 1E 7D 94 2B A8 54 15 E0 0C 8D E3 11 30 6F F7 9F 4B DB EA 84 D1 1B D8 E7 00 00 2C 00 FF C0 2C C0 2B C0 24 C0 23 C0 0A C0 09 C0 08 C0 30 C0 2F C0 28 C0 27 C0 14 C0 13 C0 12 00 9D 00 9C 00 3D 00 3C 00 35 00 2F 00 0A 01 00 00 4B 00 00 00 0E 00 0C 00 00 09 31 32 37 2E 30 2E 30 2E 31 00 0A 00 08 00 06 00 17 00 18 00 19 00 0B 00 02 01 00 00 0D 00 12 00 10 04 01 02 01 05 01 06 01 04 03 02 03 05 03 06 03 00 05 00 05 01 00 00 00 00 00 12 00 00 00 17 00 00 
 [sþ`.}+¨Tàã0o÷KÛêÑØç,ÿÀ,À+À$À#À
À	ÀÀ0À/À(À'ÀÀÀ=<5/
K	127.0.0.1


[0.540s] [LoggingConnectionAdapter] [Debug] WriteAsync[1215] 16 03 03 00 59 02 00 00 55 03 03 5B 73 17 FE 93 4A F3 88 E6 81 E2 3E E4 59 AC 25 C1 32 88 9C 2E E6 1C 08 7F CB 89 CE D4 30 4C D7 20 3B D8 FB F1 BE E0 B3 8F 5F FE EE 77 B0 6B 44 64 01 90 0F 0B BB A5 BD 3F B1 D3 9F 59 95 88 9A 2F C0 30 00 00 0D 00 00 00 00 FF 01 00 01 00 00 17 00 00 16 03 03 03 01 0B 00 02 FD 00 02 FA 00 02 F7 30 82 02 F3 30 82 01 DF A0 03 02 01 02 02 10 32 B6 9D 2C 9F C5 0F AF 4F 1D A2 21 1D 5F 70 E1 30 09 06 05 2B 0E 03 02 1D 05 00 30 14 31 12 30 10 06 03 55 04 03 13 09 6C 6F 63 61 6C 68 6F 73 74 30 1E 17 0D 31 35 30 39 31 36 30 36 32 31 30 39 5A 17 0D 31 36 30 39 31 36 30 36 32 31 30 38 5A 30 14 31 12 30 10 06 03 55 04 03 13 09 6C 6F 63 61 6C 68 6F 73 74 30 82 01 22 30 0D 06 09 2A 86 48 86 F7 0D 01 01 01 05 00 03 82 01 0F 00 30 82 01 0A 02 82 01 01 00 9A 51 99 A1 A0 D6 08 E5 D9 58 35 E5 C1 A5 09 8C 6B DF 76 95 29 C5 87 68 1B 18 87 06 25 28 C2 BF E5 F9 F8 C5 AB EC 69 AA 08 D2 FF 95 52 5D DE B8 AA AD C4 AE 87 CE 73 BC 94 AD 31 40 34 C7 59 1B A7 55 7F 34 5A A0 73 57 39 94 23 9A F6 C6 30 E3 8E 25 0C 38 CD CF C1 B7 90 D3 AB 27 91 B7 8E B6 83 12 CF FC 79 08 13 8E 4D 13 CA 91 BE C1 FA B9 4D 41 37 5A BF A4 7C 6C 9E 46 9E 51 1A 85 70 D8 15 47 57 FB 05 07 E8 6E 3D 61 A6 77 84 04 86 21 EF 91 91 A5 25 28 D3 65 52 E6 01 83 73 E0 7F B1 60 76 7E 7B 3B 9B 4E 6B 9E AB 98 9A 90 10 45 0E CD CA FA 15 9F 72 58 8F F1 05 37 B4 69 76 5B 6A 0F 2B 6F 78 C5 79 D1 24 AA 2E 72 9B 70 49 B3 06 54 59 58 B5 DF 74 53 49 8C 8C 0C 06 F2 38 CF D9 C6 90 91 4C BE 69 F4 A8 50 2E 3C 7F B9 58 46 57 37 91 7A 39 4F 37 11 0D 28 01 44 5F 9C 99 04 73 02 03 01 00 01 A3 49 30 47 30 45 06 03 55 1D 01 04 3E 30 3C 80 10 19 CE 2D 00 8B 84 C6 D5 5B C8 72 37 E4 BF 72 88 A1 16 30 14 31 12 30 10 06 03 55 04 03 13 09 6C 6F 63 61 6C 68 6F 73 74 82 10 32 B6 9D 2C 9F C5 0F AF 4F 1D A2 21 1D 5F 70 E1 30 09 06 05 2B 0E 03 02 1D 05 00 03 82 01 01 00 71 B2 95 4A 79 28 61 C7 22 57 C0 76 68 1C 19 9F 3D C6 D1 39 86 93 F4 5A C7 7C 2A 51 3E 1C A7 C8 0D BB 79 48 1E D8 04 69 7C 80 43 14 00 63 27 56 F0 64 44 7B 4B 99 32 4E D0 4C F7 54 E9 93 2C 20 0E CB 03 A4 76 E1 57 36 C3 A3 56 4C 0F C8 25 27 2E 5C FB 12 B8 8D A4 D9 8F 4F D6 56 4D 54 1D 15 5E C5 45 78 63 12 04 57 2C 25 FE B4 7C F5 20 5E BB EC C2 31 EA 42 F8 EC 09 95 2A 7A FE F6 0F 85 AF 90 5F 64 39 0D 66 96 A9 24 AD E6 7E 52 1D 68 CD 78 8C EC FA AE CA AC 5A C2 4C 08 E9 50 3C A7 25 B6 AE C1 4C A8 08 AA E3 4F 0D E8 47 C8 54 23 6C CB CC 66 0F 4A 7F F6 1B A6 E2 4A 16 90 B3 7E 2F AE 88 CC 9C E5 2B 41 45 52 05 D9 BB 8A 42 7B C2 8A E4 DE E8 75 04 E9 97 2B 27 1B 04 C7 B7 52 27 F9 46 E6 1D E0 3E C7 E8 35 6C 67 4E 2A 26 59 5F D9 A7 61 36 63 07 DA BE 33 E1 5A 3A C1 C2 B2 16 03 03 01 4D 0C 00 01 49 03 00 17 41 04 0E 66 48 17 81 54 6F AF 1B 4B D1 A8 F8 0D 61 5D EA 74 85 F9 A2 D9 BE A4 3D CC 4F A6 C9 55 BF 82 23 43 9E 37 9A E3 6F 89 28 BA 51 82 CB 9A B1 F2 6B 73 2A 0E 6C D8 2A 72 9F B8 80 04 5A 3A 29 1D 04 01 01 00 04 D6 9E CA CE E3 A5 19 35 99 D7 B4 E0 C9 BD 18 03 3F 69 B4 DD E2 DD FE 6F 9A 2D 7C 01 27 12 88 AC 87 FB 78 4E 91 F2 C3 F5 DC DA 32 9B 3A D6 13 46 6C 1C A4 39 88 76 40 01 68 4C B1 62 F2 05 90 6A 47 A8 EA 13 9B B7 A8 D6 56 AB 74 62 4D E0 60 E2 C6 EA A0 6A 84 65 9D C4 0C 74 87 6D 8A C9 8D B0 22 6B A5 82 A0 7A 1D E9 8F B0 B8 82 80 07 79 9E 58 00 7A 95 75 CB 04 0E 36 AA 31 67 2E D0 17 94 43 5B 74 D8 31 02 4D A3 21 EA 24 C4 78 0A E6 06 BB C8 79 DD 33 93 C6 7E 43 8A DA 70 A4 34 E3 42 D2 9A 08 FC AC 9F 0F BD 4D 99 27 81 61 9D A4 0D 46 19 8D E4 03 2E 6C 07 18 05 2C D4 04 F6 D5 75 37 C8 E5 C2 1F F6 6E 8E B9 26 E4 7E C5 18 F8 6B F0 7C 2F B4 C3 30 DA 87 A0 B6 48 CD 1B C1 CB AD 99 D7 36 64 46 2F B5 6C FA 8F A6 25 C2 BC 9C E2 56 35 56 EC 76 7A 80 71 64 76 49 5E CF 98 30 16 03 03 00 04 0E 00 00 00 
YU[sþJóæâ>äY¬%Á2.æËÎÔ0L× ;Øûñ¾à³_þîw°kDd»¥½?±ÓY/À0
ÿýú÷0ó0ß 2¶,Å¯O¢!_pá0	+010U	localhost0
150916062109Z
160916062108Z010U	localhost0"0
	*H÷
0
Q¡ ÖåÙX5åÁ¥	kßv)Åh%(Â¿åùøÅ«ìiªÒÿR]Þ¸ª­Ä®Îs¼­1@4ÇY§U4Z sW9#öÆ0ã%8ÍÏÁ·Ó«'·¶ÏüyMÊ¾Áú¹MA7Z¿¤|lFQpØGWûèn=a¦w!ï¥%(ÓeRæsà±`v~{;Nk«EÍÊúrXñ7´iv[j+oxÅyÑ$ª.rpI³TYXµßtSIò8ÏÙÆL¾iô¨P.<¹XFW7z9O7
(D_s£I0G0EU>0<Î-ÆÕ[Èr7ä¿r¡010U	localhost2¶,Å¯O¢!_pá0	+q²Jy(aÇ"WÀvh=ÆÑ9ôZÇ|*Q>§È
»yHØi|Cc'VðdD{K2NÐL÷Té, Ë¤váW6Ã£VLÈ%'.\û¸¤ÙOÖVMT^ÅExcW,%þ´|õ ^»ìÂ1êBøì	*zþö¯_d9
f©$­æ~RhÍxìú®Ê¬ZÂLéP<§%¶®ÁL¨ªãO
èGÈT#lËÌfJö¦âJ³~/®Ìå+AERÙ»B{ÂäÞèué+'Ç·R'ùFæà>Çè5lgN*&Y_Ù§a6cÚ¾3áZ:ÁÂ²MIAfHTo¯KÑ¨ø
a]êtù¢Ù¾¤=ÌO¦ÉU¿#C7ão(ºQË±òks*lØ*r¸Z:)ÖÊÎã¥5×´àÉ½?i´ÝâÝþo-|'¬ûxNòÃõÜÚ2:ÖFl¤9v@hL±bòjG¨ê·¨ÖV«tbMà`âÆê jeÄtmÉ°"k¥ zé°¸yXzuË6ª1g.ÐC[tØ1M£!ê$Äx
æ»ÈyÝ3Æ~CÚp¤4ãBÒü¬½M'a¤
Fä.l,ÔöÕu7ÈåÂön¹&ä~Åøkð|/´Ã0Ú ¶HÍÁË­×6dF/µlú¦%Â¼âV5VìvzqdvI^Ï0
[0.553s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
[0.554s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Receiving data with a buffer at 4504 with count 4096
[0.554s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Received 126 bytes of data
[0.554s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.555s] [LoggingConnectionAdapter] [Debug] ReadAsync[5] 16 03 03 00 46 
F
[0.555s] [LoggingConnectionAdapter] [Debug] ReadAsync[70] 10 00 00 42 41 04 45 D9 15 27 17 FA 58 DD FF 77 03 99 0F 7A C5 FA 81 26 B7 D5 37 33 38 F4 6F 02 8B A4 77 AB 96 03 AC 81 1D DB 2F 6E 8C 25 51 BA B4 53 F1 7A 50 B1 F1 CF C0 7B 25 69 78 5C 09 15 FF 54 C6 24 79 F9 
BAEÙ'úXÝÿwzÅú&·Õ738ôo¤w«¬Û/n%Qº´SñzP±ñÏÀ{%ix\	ÿTÆ$yù
[0.557s] [LoggingConnectionAdapter] [Debug] ReadAsync[5] 14 03 03 00 01 

[0.557s] [LoggingConnectionAdapter] [Debug] ReadAsync[1] 01 

[0.558s] [LoggingConnectionAdapter] [Debug] ReadAsync[5] 16 03 03 00 28 
(
[0.559s] [LoggingConnectionAdapter] [Debug] ReadAsync[40] F6 8C 3F A6 E1 1C 8B 3D 27 78 70 68 8B 31 1C BD C6 7E 12 4B 1F 8E F4 EF 69 80 25 0E 1B C6 65 40 37 66 11 EF 79 A4 66 E8 
ö?¦á='xph1½Æ~Kôïi%Æe@7fïy¤fè
[0.559s] [LoggingConnectionAdapter] [Debug] WriteAsync[51] 14 03 03 00 01 01 16 03 03 00 28 77 69 09 B1 51 86 FA 12 FD FB 3F C1 AC 30 8B 0C 49 97 17 53 44 D7 7A B2 9B 7F 97 10 77 A6 FC 3C 48 6C 1A A7 D9 E2 0A 68 
(wi	±Qúýû?Á¬0ISD×z²w¦ü<Hl§Ùâ
h
[0.583s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
[0.583s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Receiving data with a buffer at 16792 with count 4096
[0.583s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Received 70 bytes of data
...
```
But what I actually see is:
```
[0.001s] [TestLifetime] [Information] Starting test "DoesNotOverrideDirectConfigurationWithIServerAddressesFeature_IfPreferHostingUrlsFalse" at "2018-08-14T03:25:23"
[0.005s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting starting
[0.086s] [Microsoft.AspNetCore.Server.Kestrel] [Warning] Overriding address(es) '"http://127.0.0.1:0"'. Binding to endpoints defined in "UseKestrel()" instead.
[0.086s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting started
[0.087s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Loaded hosting startup assembly "Sockets.BindTests"
[0.131s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLG1IM80TNFQ"" started.
[0.133s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.133s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
[0.134s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Receiving data with a buffer at 125072 with count 4096
[0.134s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Received 169 bytes of data
[0.134s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.134s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
[0.135s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Receiving data with a buffer at 125241 with count 3927
[0.135s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Received 169 bytes of data
[0.135s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.135s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
[0.135s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Receiving data with a buffer at 125410 with count 3758
[0.136s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Received 169 bytes of data
[0.136s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.136s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
[0.140s] [LoggingConnectionAdapter] [Debug] ReadAsync[5] 16 03 02 00 A4 
¤
[0.141s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Receiving data with a buffer at 125579 with count 3589
[0.141s] [LoggingConnectionAdapter] [Debug] ReadAsync[164] 01 00 00 A0 03 03 5B 72 4B A3 81 82 82 A3 2A 3F 0D 2B 18 BC 5A 32 57 90 38 0A 7E 4B 06 A2 94 92 C1 B1 49 4E 54 9C 00 00 2C 00 FF C0 2C C0 2B C0 24 C0 23 C0 0A C0 09 C0 08 C0 30 C0 2F C0 28 C0 27 C0 14 C0 13 C0 12 00 9D 00 9C 00 3D 00 3C 00 35 00 2F 00 0A 01 00 00 4B 00 00 00 0E 00 0C 00 00 09 31 32 37 2E 30 2E 30 2E 31 00 0A 00 08 00 06 00 17 00 18 00 19 00 0B 00 02 01 00 00 0D 00 12 00 10 04 01 02 01 05 01 06 01 04 03 02 03 05 03 06 03 00 05 00 05 01 00 00 00 00 00 12 00 00 00 17 00 00 
 [rK££*?
+¼Z2W8
~K¢Á±INT,ÿÀ,À+À$À#À
À	ÀÀ0À/À(À'ÀÀÀ=<5/
K	127.0.0.1


[0.141s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Received 169 bytes of data
[0.141s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waiting for data
[0.142s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Waited for 0 bytes of data
...
```

Any ideas why sockets may be receiving the same data over and over again? Note that this only affects macOS and does not repro on Windows and Linux as far as I can tell. Also there appears to be a race since it does not repro all the time. 

## To repro:
1. Clone the working branch: https://github.com/aspnet/KestrelHttpServer/tree/tmds/zero-byte-read
2. Run the test DoesNotOverrideDirectConfigurationWithIServerAddressesFeature_IfPreferHostingUrlsFalse at /test/Kestrel.Transport.Sockets.BindTests: `dotnet test --filter DoesNotOverrideDirectConfigurationWithIServerAddressesFeature_IfPreferHostingUrlsFalse`

cc @halter73 @Tratcher

**Comment by JunTaoLuo** [comment]
cc @karelz @geoffkizer

**Comment by karelz** [comment]
Are you able to reproduce it outside of Kestrel? (to remove additional moving pieces)

**Comment by JunTaoLuo** [comment]
I have not yet tried to reproduce this on its own yet but we suspect this is a race in Sockets issue since the same implementation works consistently on Windows and Linux and the failure on macOS is intermittent as the bug seems to manifest only about half the time.

**Comment by stephentoub** [comment]
> I have not yet tried to reproduce this on its own yet 

@JunTaoLuo, any progress on a repro?

**Comment by JunTaoLuo** [comment]
I added a repro at https://github.com/JunTaoLuo/ZeroByteRead. To repro, run the app until the `Waiting to accept` message shows up, then navigate to http://localhost:5000 in your browser (or send a request via curl). If you run on mac, you'll see the following logs repeat over and over again:
```
Waiting for data
Waited for 0 bytes of data
Received 464 bytes of data
GET / HTTP/1.1
Host: localhost:5000
Connection: keep-alive
Cache-Control: max-age=0
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8,zh-CN;q=0.7,zh-TW;q=0.6,zh;q=0.5
```

If you run this on Linux or if you comment out the zero byte reads lines (https://github.com/JunTaoLuo/ZeroByteRead/blob/master/Program.cs#L142-L143), you'll see that you don't run into infinite repeated reads.

**Comment by stephentoub** [comment]
Any chance you have a repro without ASP.NET so we can better isolate it?

**Comment by stephentoub** [comment]
Oh, nevermind my question... this repro is without ASP.NET.  Thanks.

**Comment by wfurt** [comment]
I run it under dtrace and the recvmsg() is always called with MSG_PEEK (0x2)
> 18711/0xd7af6:  recvmsg(0x24, 0x70000538EBA8, 0x2)		 = 78 0

I put in quick debug to TryCompleteReceiveFrom() and the Peek flags is passed in even with valid buffer size:
```
Waiting for data
TryCompleteReceiveFrom called with length=0 and flags Peek
Waited for 0 bytes of data
TryCompleteReceiveFrom called with length=2048 and flags Peek
Received 78 bytes of data
```

**Comment by stephentoub** [comment]
We pass Peek when doing a 0-byte read here:
https://github.com/dotnet/corefx/blob/a1db302fc7b876675618d9aa4d6d53989510b97c/src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L603-L610
I'm not sure how you're getting Peek with a non-zero length, though.

**Comment by wfurt** [comment]
When the Receive() finishes, receivedFlags is set to Peek on OSX (set to None on Linux) 
That seems like platform difference where OSX propagates passed Peek flag to  receivedFlags bitmask. 

and than for example  in ReceiveMessageOverlappedAsyncResult.Unix.cs we do ```_socketFlags = receivedFlags;``` 

so when the "zero" read finishes _awaitableEventArgs.SocketFlags in the repro code has Peek set. 
When called again for "normal" read, it only peeks message. 

Following change in repro code seems to fix the observed issue:
```
--- a/Program.cs
+++ b/Program.cs
@@ -97,12 +97,16 @@ public SocketAwaitableEventArgs WaitForDataAsync()
         public SocketAwaitableEventArgs ReceiveAsync(Memory<byte> buffer)
         {
             _awaitableEventArgs.SetBuffer(buffer);
+            _awaitableEventArgs.SocketFlags = SocketFlags.None;

             if (!_socket.ReceiveAsync(_awaitableEventArgs))
             {
                 _awaitableEventArgs.Complete();
             }
```

does it make sense? This is either issue with reusing SocketAsyncEventArgs or we should filter out the flag when we set it behind the scenes.

**Comment by stephentoub** [comment]
> That seems like platform difference where OSX propagates passed Peek flag to receivedFlags bitmask.

Nice sleuthing.  It's unfortunate that macOS has such a difference, but that makes sense and explains why our test for this (https://github.com/dotnet/corefx/blob/8bc7dbe79b16c3f5823871d9664555a69629c3e6/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs#L641) didn't catch it.  The test needs to be made tighter in what it validates.

> This is either issue with reusing SocketAsyncEventArgs or we should filter out the flag when we set it behind the scenes.

Is there ever a case where the OS would legitimately set peek into the result flags?  If not, I think we should filter it out and that should address the issue.  If there's a legitimate case for it, the problem becomes harder.

**Comment by wfurt** [comment]
no, I don't think so. I'll do more testing and I'll put fix together. I also like if we filter it out so the socket surface is more consistent and hides platform differences. I was planning to give it quick try on FreeBSD but I did not make it today.

**Comment by JunTaoLuo** [comment]
Any chance this can make it into 2.2?

cc @muratg

**Comment by karelz** [comment]
@JunTaoLuo @muratg 2.2 is only per request based on business impact - what is it blocking?

**Comment by JunTaoLuo** [comment]
Ideally, we would like to make Kestrel's behaviour the same across all platforms. Currently, we do 0 byte reads to reduce memory usage on linux and windows but we were blocked on mac due to this issue. I was under the impression that this would be a small fix so I was wondering if it was possible to do it in 2.2.

**Comment by karelz** [comment]
That sounds more like "very nice to have" rather than "blocking".
Given that we do not have a fix even in master yet, it may be too late to ask for it to be in 2.2 at this point.
